PDF Generation for Query Reports

javierwonghf opened the pull request:
- Added option in reports to be published to PDF
- This uses existing frappe functionalities to generate the HTML content together with the letterhead and styling and pushes the HTML content to an API which in turn generates the PDF to be sent back to the client
- The only snag is that I can't seem to manipulate the saving filename when a user saves the PDF file, and because a blob is used, the proposed filename (which can be changed in the save dialog box) in Chrome becomes a randomized string as opposed to something user readable. Firefox merely proposes "document.pdf" as the filename which I think is appropriate

javierwonghf committed:
Merge pull request #2 from frappe/develop

13/012016 Merge

javierwonghf committed:
Added Report_To_PDF

For printing of reports in PDF using a HTML stream

javierwonghf committed:
Added Generation of PDF for Query Reports

Added option to generate PDF from Query Reports. Supports both reports with HTML templates and those that are grids only.

rmehta commented on frappe/public/js/frappe/views/reports/query_report.js:
```diff
@@ -135,6 +137,47 @@ frappe.views.QueryReport = Class.extend({
+		formData.append("html", html);
+		var blob = new Blob([], { type: "text/xml"});
+		//formData.append("webmasterfile", blob);
+		formData.append("blob", blob);
```
do we need to send this blob?

bohlian replied:
Good question, I did try it without the blob but the API would not process the form data via POST if without. I'm not sure why, perhaps it was needed to explicitly define the type i.e. text/xml. Either way I would have ideally wanted to just a simple AJAX call but the APIs doesnt seem to like accepting params if the AJAX is POST. Tried troubleshooting down all the way till response.py and I couldn't get further than that. I suspect it may have to be due to a very low level python response behavior (could be wrong though).

rmehta commented on frappe/public/js/frappe/views/reports/query_report.js:
```diff
@@ -135,6 +137,47 @@ frappe.views.QueryReport = Class.extend({
+		xhr.open("POST", '/api/method/frappe.templates.pages.print.report_to_pdf');
+		xhr.setRequestHeader("X-Frappe-CSRF-Token", frappe.csrf_token);
+		xhr.responseType = "arraybuffer";
+
```
why not use `frappe.call` ?

bohlian replied:
Dont think its possible, I need to return an arraybuffer to generate the PDF on the fly. Perhaps you have an idea of how we can use frappe.call?

anandpdoshi commented:
@bohlian this seems like a workaround. Ideally it should just generate PDF from server side without the intervention of the browser just like generating pdf of a doc

bohlian commented:
@anandpdoshi true, I would have preferred that as well, the print architecture seems different for docs (server side) than for reports (client side). Not sure why it was built that way. I supposed I didn't want to disturb too much of the existing structure.

What do you guys think? Perhaps use this first and we can redesign later?

anandpdoshi commented:
As a workaround, one can always use the browser’s print functionality to output as PDF. It is best to fix this at its core like we did for docs printing. 

For that we will need to refactor the query_report.py to evaluate the columns and data into a list of dicts and pass them as context to a ninja template. This will require porting the report print microtemplate to jinja. 

https://github.com/frappe/frappe/blob/develop/frappe/public/js/frappe/views/reports/print_grid.html

Best,
Anand Doshi

ERPNext

> On 14-Jan-2016, at 1:12 PM, Javier Wong <notifications@github.com> wrote:
> 
> @anandpdoshi <https://github.com/anandpdoshi> true, I would have preferred that as well, the print architecture seems different for docs (server side) than for reports (client side). Not sure why it was built that way. I supposed I didn't want to disturb too much of the existing structure.
> 
> What do you guys think? Perhaps use this first and we can redesign later?
> 
> —
> Reply to this email directly or view it on GitHub <https://github.com/frappe/frappe/pull/1517#issuecomment-171564170>.
>

rmehta commented:
@bohlian our internal assessment is that most browsers already have a print to pdf feature built-in. What you are doing is sending an HTML blob to server and then just dumping back the PDF. Do you have a strong use case for this?

ccfiel commented:
@rmehta  I think the strong use case for this is html is limited in formatting the report example how can I used this Header/Footer? Its just limited for pdf generation. https://discuss.erpnext.com/t/header-footer-development/7442/57.

bohlian commented:
@rmehta my points as follows:
1) When you print from a browser it typically contains the Title of the page on the top together with the date on the top left. This is automatically inserted by the browser and is not part of the print format. This is quite undesireable.
2) Printing PDF using a browser also requires a 3 step process. First, you have to click Menu -> Print (In the Report), then on the browser  you have to select File -> Print, then select Print to PDF, then type in a file name then save.
3) If we implemented this solution it would just be Menu -> Print (In the Report), then click Save.
4) I do however agree with @anandpdoshi that it would be alot better if we could implement server side PDF generation which I am working but without an ETA at the moment largely because it is fairly complex. As such, I asked if we could use this as a temporary solution until we get that part developed.

What are your thoughts?

ccfiel commented:
Hello @bohlian @anandpdoshi @rmehta , 

How much would it cost for the development for this to be optimized and put it in the core of frappe? I would like sponsor :) and maybe other can also co-sponsor this. 

Chris

javierwonghf committed:
Merge pull request #3 from frappe/develop

1/3/2016 Merge

javierwonghf committed:
[fixed] Multiple Sessions are Always Denied

Casted frappe.db.get_system_setting('deny_multiple_sessions') to int, as a string it is always evaluated as True in the conditional statement and therefore it is always treated as multiple sessions are denied.

javierwonghf committed:
Revert "[fixed] Multiple Sessions are Always Denied"

This reverts commit df085e56ed44b63ca6f74a8f6d108ebb61783bca.

rmehta commented on frappe/public/js/frappe/views/reports/query_report.js:
```diff
@@ -135,6 +137,47 @@ frappe.views.QueryReport = Class.extend({
+			//Render Report in HTML
+			var html = frappe.render_template("print_template", {content:content, title:__(this.report_name)});
+		}
+
```
could have avoided duplicating this section.

Merging manually for now.

ccfiel commented:
@rmehta thanks a lot!
